explainer.md

```diff
@@ -345,6 +341,114 @@ And here's the result:
+
+#### Script on Old Document
+```js
+document.addEventListener("viewTransitionOnOldDocument", (event) => {
+  // Cancel the transition (based on new URL) if needed.
+  if (shouldNotTransition(event.url)) {
+    event.preventDefault();
+    return;
+  }
+    
+  // Set up names on elements based on the new URL.
+  if (shouldTagThumbnail(event.url))
+    thumbnail.style.viewTransitionName = "full-embed";
+    
+  // Add opaque contextual information to share with the new Document.
+  // This must be [serializable object](https://developer.mozilla.org/en-US/docs/Glossary/Serializable_object).
+  event.context = createTransitionContext(event.url);
+});
+```
+
+#### Script on New Document
+```js
+// This event must be registered before the `body` element is parsed.
+document.addEventListener("viewTransitionOnNewDocument", (event) => {
+  // Cancel the transition (based on old URL) if needed.
+  if (shouldNotTransition(event.url)) {
+    event.preventDefault();
+    return;
+  }
+  
+  // The `ViewTransition` object associated with this transition.
+  const transition = event.transition;
+  
+  // Retrieve the list of names captured on the old Document.
+  const oldNames = event.oldNames;
+  
+  // Retrieve the context generated by the old Document.
+  const context = event.context;
+  
+  // Add render-blocking resources to delay the first paint and transition
+  // start. This can be customized based on the old Document state when the
+  // transition was initiated.
+  markRenderBlockingResources(oldNames, context);
+  
+  // The browser defers painting the new Document until all render-blocked
+  // resources have been fetched or timed out.
+  // The `domUpdated` promise can be used to inspect when the above is
+  // done and add viewTransitionNames based on what the new Document's
+  // first paint would look like.
+  transition.domUpdated.then(() => {
+     thumbnail.style.viewTransitionName = "full-embed";
+  });
+  
+  // The `ready` promise resolves when the pseudo-elements have been generated
+  // and can be used to customize animations via script.
+  transition.ready.then(() => {
+    document.documentElement.animate(...,
+       {
+         // Specify which pseudo-element to animate
+         pseudoElement: "::view-transition-new(root)",
+       }
+    );
+  });
+  
+  // The `finished` promise resolves when all animations have been executed.
+  transition.finished.then(() => {
+    // Remove viewTransitionNames tied to this transition.
+    thumbnail.style.viewTransitionName = "none";
+  });
+});
+```
+
+This provides the same scripting points as the SPA API, allowing developers to set class names to tailor the animation to a particular type of navigation.
+
+Issue: Do we need better timing for viewTransitionOnNewDocument event? Especially for Documents restored from BFCache.
```

Comment:
Fwiw, I like `beforepagehide` (or maybe just `pagehide`??) and `beforepageshow`.

Reply:
Do we need a generic event for this? I was hoping for something ViewTransition specific so we can skip dispatching if there is no developer opt-in to the API. Don't want to add more events that must be dispatched for all navigations (especially the back/fwd ones which can be completely browser driven).

We'll also not dispatch these events if the browser decided to skip the transition for any reason.

Reply:
Yeah, I'm coming round to the specific event idea.

Can't help myself bikeshedding:

- `documenttransitionoldcapture` - fired on the outgoing page. Can't think of anything shorter.
- `documenttransition` - fired on the new page.

It feels like "document transition" is right here, to differentiate it from SPA transitions.